juju/model.py

```diff
@@ -771,10 +771,14 @@ async def watch_received_waiter():
                 raise JujuError("AllWatcher task is finished abruptly without an exception.")
             raise self._watcher_task.exception()
 
-        if self.info is None:
-            contr = await self.get_controller()
-            self._info = await contr.get_model_info(model_name, model_uuid)
-            log.debug('Got ModelInfo: %s', vars(self.info))
+        if self._info is None:
+            # TODO (cderici): See if this can be optimized away, or at least
+            # be done lazily (i.e. not everytime after_connect, but whenever
+            # self.info is needed -- which here can be bypassed if model_uuid
+            # is known)
+            async with ConnectedController(self.connection()) as contr:
+                self._info = await contr.get_model_info(model_name, model_uuid)
```

Comment:
https://pypi.org/project/async-property/

Supposedly has things like:
`@async_cached_property`
Which will call the function only once and otherwise cache the result.

That said, I don't think async and properties act nice together, because then you need to 'foo = await obj.foo' which really doesn't feel good (I generally don't think accessing a property should trigger an HTTP call.)

Even if their doc, this doesn't feel good:
```
>>> instance.value
<AwaitableOnly "Foo.value">

>>> await instance.value
loading value
123
>>> await instance.value
123
>>> instance.value
123
```
The fact that you don't always have to await, but sometimes you do is really not a great UX.

So I was wrong to ask you to do so as part of the property, and checking _info is ok.